client/vproto/plugin.go

```diff
@@ -0,0 +1,315 @@
+	pluginTimeout = 3 * time.Second
+)
+
+// plugin is the stub for a registered (configured or connected) plugin.
+//
+// Plugins are either declared statically or connected dynamically. Static
+// plugins are declared in the NRI configuration file. They are started by
+// the NRI runtime client when it gets instantiated/started by the runtime.
+// Dynamic plugins are launched by means external to NRI and they register
+// themselves by connecting to a dedicated NRI socket at a well-known path.
+//
+// Configuration for static plugins is part of the NRI configuration. Dynamic
+// plugins can acquire their configuration by some other means, external to
+// NRI.
+type plugin struct {
+	name   string            // plugin name
+	id     string            // plugin id (sort order for dynamic plugins)
+	pid    int               // plugin pid for dynamic plugins
+	cmd    *exec.Cmd         // plugin Cmd (if exec()ed by us)
+	stdin  io.WriteCloser    // plugin stdin (if exec()ed by us)
+	conn   net.Conn          // plugin connection
+	ttrpcc *ttrpc.Client     // plugin ttRPC client
+	client api.PluginService // plugin service
+	events int               // plugin event subscription
+	closed bool              // plugin has closed connection
+}
+
+// close the connection to the plugin.
+func (p *plugin) close() error {
+	if p.ttrpcc != nil {
+		if err := p.ttrpcc.Close(); err != nil {
+			if err != ttrpc.ErrClosed {
+				return err
+			}
+		}
+		p.ttrpcc = nil
+	}
+	if p.conn != nil {
+		if err := p.conn.Close(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// stop the plugin.
+func (p *plugin) stop() {
+	if p.cmd != nil && p.cmd.Process != nil {
+		p.cmd.Process.Kill()
+		p.cmd.Process.Wait()
+		p.cmd.Process.Release()
+	}
+}
+
+// configure the plugin and subscribe it for the events it requests.
+func (p *plugin) configure(ctx context.Context, config string) error {
+	ctx, cancel := context.WithTimeout(ctx, pluginTimeout)
+	defer cancel()
+
+	rpl, err := p.client.Configure(ctx, &api.ConfigureRequest{
+		Config: config,
+	})
+	if err != nil {
+		return errors.Wrap(err, "failed to configure plugin")
+	}
+
+	if rpl.Id != "" {
+		p.id = rpl.Id
+	} else {
+		p.id = p.name
+	}
+
+	if p.cmd == nil && p.id == "" {
+		return errors.New("dynamic plugin failed to identify itself")
+	}
+
+	eventMask := map[api.Event]int{
+		api.Event_RUN_POD_SANDBOX:    runPodSandbox,
+		api.Event_STOP_POD_SANDBOX:   stopPodSandbox,
+		api.Event_REMOVE_POD_SANDBOX: removePodSandbox,
+		api.Event_CREATE_CONTAINER:   createContainer,
+		api.Event_START_CONTAINER:    startContainer,
+		api.Event_UPDATE_CONTAINER:   updateContainer,
+		api.Event_STOP_CONTAINER:     stopContainer,
+		api.Event_REMOVE_CONTAINER:   removeContainer,
+		api.Event_ALL:                allEvents,
+	}
+
+	for _, event := range rpl.Subscribe {
+		mask, ok := eventMask[event]
+		if !ok {
+			return errors.Errorf("invalid plugin event: %v", event)
+		}
+		p.events |= mask
+	}
+
+	if p.events == 0 {
```

Comment:
Hmm, I see your point and I think that'd make sense. That would provide a way for the plugin to request a graceful shutdown based on its configuration.

But considering that this is not a bitmask at the protocol level (because of the poor/nonexistent support in protobuf for that) but rather an array/slice of events names, I think it would be better to add an explicit reserved subscription identifier for that. For instance `none` to complement the already existing `all`. That would have the additional benefit that `0` could continue to be special-cased to mean `all`, so the default zero-value protocol message would still result in a subscription to all protocol messages instead of being shut down.

Reply:
Hmm, then again your point below about backward-compatible protocol changes made me doubt if `all` makes sense at all...